### ISSUE TITLE
Getting predictions and gold labels from `trainer.test()`

### ISSUE THREAD

**Issue by ZhaofengWu** [PyTorchLightning/pytorch-lightning#12969]
### Discussed in https://github.com/PyTorchLightning/pytorch-lightning/discussions/12841

<div type='discussions-op-text'>

<sup>Originally posted by **ZhaofengWu** April 21, 2022</sup>
Is there a way to obtain predictions and gold labels from `trainer.test()`? I have access to these in `test_epoch_end()`, but since this function doesn't return, I can't pass them outside. Options that I've considered:

1. Use `self.log("preds", preds)`, but apparently torch tensors are automatically reduced.
2. Use `trainer.predict()`, but since I'm already doing `trainer.test()`, having to run the forward pass on the evaluation data twice seems wasteful.
3. Set `self._some_custom_property_{preds,labels}` in `test_epoch_end()`, and access these properties outside. I'd go for this if there's no better alternative, but it sounds like a hack.</div>

This discussion topic might have been overlooked -- I think it can also be viewed as sort of a feature request, if there isn't currently an easy way to achieve this, so I hope it's ok to convert it into an issue.

cc @awaelchli @rohitgr7

**Comment by rohitgr7** [comment]
unfortunately `test`is used to evaluate your model using metrics only and can't be used to return predictions. But you can use a callback to store them.

```py
class TestPredStore(Callback):
    def __init__(self):
        self.test_predictions = []

    def on_test_batch_end(self, trainer, pl_module, outputs, *args, **kwargs):
        self.test_predictions.append(outputs)  # or extend

trainer = Trainer(callbacks=[TestPredStore()])
trainer.test(...)
```

**Comment by ZhaofengWu** [comment]
Thanks!
